// @global-genius/identity/application/use-cases/login.ts

import { IPasswordHasher } from "../../domain/services/ipassword-hasher";
import { ITokenService } from "../../domain/services/itoken";
import { InvalidCredentialsError } from "../../domain/errors/invalid-credentials";
import { Session } from "../../domain/entities/session";
import { LoginUserResponseDTO } from "../../domain/dtos/auth-response";
import { LoginUserInputDTO } from "../../domain/dtos/login";
import { ILoginRepository, ISessionRepository } from "@global-genius/database";
import { IUseCase } from "../../domain/usecases/iusecase";

export class LoginUserUseCase implements IUseCase<LoginUserInputDTO,LoginUserResponseDTO>{
  constructor(
    private readonly loginRepository: ILoginRepository,
    private readonly sessionRepository: ISessionRepository,
    private readonly passwordHasher: IPasswordHasher,
    private readonly tokenService: ITokenService,
  ) {}

  async execute(input: LoginUserInputDTO): Promise<LoginUserResponseDTO> {
    const user = await this.loginRepository.findByEmail(input.email);

    if (!user) {
      throw new InvalidCredentialsError();
    }

    const passwordMatches = await this.passwordHasher.compare(
      input.password,
      user.password,
    );

    if (!passwordMatches) {
      throw new InvalidCredentialsError();
    }

    const accessToken = this.tokenService.generateAccessToken(
      user.id as string,
    );

    const refreshToken = this.tokenService.generateRefreshToken(
      user.id as string,
    );

    const session = new Session(
      user.id as string,
      refreshToken,
      new Date(),
      new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
    );

    await this.sessionRepository.create(session);

    return {
      accessToken,
      refreshToken,
      user: {
        id: user.id as string,
        email: user.email,
        role: user.role,
      },
    };
  }
}
