// @global-genius/identity/infrastructure/repositories/in-memory/user.ts

import { ILoginRepository, IRegisterRepository, IUserRepository, User } from "@global-genius/database";
import { CreateUserInput } from "@global-genius/database/src/repositories/types/create-user";
import { User as UserEntity } from "../../../domain/entities/user";

export class InMemoryUserRepository implements IUserRepository,ILoginRepository,IRegisterRepository {
  async update(id:string,data: any): Promise<User> {
    const user = this.users.find(x => x.id === id);
    user && user.changeRole(data.role);

    return user as User;
  }
  delete(id: string): Promise<User> {
    throw new Error("Method not implemented.");
  }
  private users: UserEntity[] = [];
    async findByEmail(email: string): Promise<User | null> {
    return this.users.find(u => u.email === email) || null;
  }

  async create(user: CreateUserInput): Promise<User> {
    const userEntity = UserEntity.create({email:user.email,password:user.password})
    this.users.push(userEntity);

    return userEntity;
  }



  async findById(id: string): Promise<User | null> {
    return this.users.find(u => u.id === id) || null;
  }

  deteteUsers(){
    this.users = []
  }

  async getAllUsers():Promise<User[]>{

    return this.users;
  }
}
